frontend/src/main/scala/bloop/bsp/BloopBspServices.scala

```diff
@@ -930,30 +930,30 @@ final class BloopBspServices(
         projects: Seq[ProjectMapping],
         state: State
     ): BspResult[bsp.SourcesResult] = {
-
-      val response = bsp.SourcesResult(
-        projects.iterator.map {
-          case (target, project) =>
-            val sources = project.sources.map {
-              s =>
-                import bsp.SourceItemKind._
-                val uri = s.underlying.toUri()
-                val (bspUri, kind) = if (s.exists) {
-                  (uri, if (s.isFile) File else Directory)
-                } else {
-                  val fileMatcher = FileSystems.getDefault.getPathMatcher("glob:*.{scala, java}")
-                  if (fileMatcher.matches(s.underlying.getFileName)) (uri, File)
-                  // If path doesn't exist and its name doesn't look like a file, assume it's a dir
-                  else (new URI(uri.toString + "/"), Directory)
-                }
-                // TODO(jvican): Don't default on false for generated, add this info to JSON fields
-                bsp.SourceItem(bsp.Uri(bspUri), kind, false)
+      val sourcesItems = projects.iterator.map {
+        case (target, project) =>
+          project.allSourceFilesAndDirectories.map { sources =>
+            val items = sources.map { s =>
+              import bsp.SourceItemKind._
+              val uri = s.underlying.toUri()
+              val (bspUri, kind) = if (s.exists) {
+                (uri, if (s.isFile) File else Directory)
+              } else {
+                val fileMatcher = FileSystems.getDefault.getPathMatcher("glob:*.{scala, java}")
```

Comment:
How are you implementing file watching for your glob expansion? Unlike literal file matches, this requires a watch on the parent directory.

Reply:
This particular method is called for the BSP “buildTarget/sources” request. File watching for globs will need to watch the “SourcesGlobs.directory” field (not implemented in this PR)